### PR TITLE
Prebuild lunr search index as .lunr.json when searchEngine is lunr

### PR DESCRIPTION
**Pull request by yufeih**
To support site search for static page using [lunr](https://lunrjs.org), docfx prebuild lunr search index into a `.lunr.json` file when `searchIndexType` is set to `lunr`.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6420)

### CODE REVIEW

File: test/docfx.Test/lib/JavascriptEngineTest.cs [hunk 1]
```diff
@@ -39,5 +39,15 @@ public void RunJavascriptError(string input)
                 Assert.Throws<JavaScriptEngineException>(() => engine.Run("index.js", "main", inputJson));
             }
         }
+
+        [Fact]
+        public void SupportModuleExports()
```
Comment (yufeih) [inline test/docfx.Test/lib/JavascriptEngineTest.cs:44]: Supports NodeJS `module.exports` standard so we can `require('./lunr.js')` properly.

File: src/docfx/build/search/SearchIndexBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            if (metadata.Robots != null && metadata.Robots.Contains("noindex", StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        [JsonObject(NamingStrategyType = typeof(CamelCaseNamingStrategy))]
+        private class LunrSearchItem
```
Comment (live1206) [inline src/docfx/build/search/SearchIndexBuilder.cs:78]: put this into a separate file

File: src/docfx/build/search/SearchIndexBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            var documents = JToken.FromObject(_searchIndex.Values);
+            var js = JavaScriptEngine.Create();
+            var scriptPath = Path.Combine(AppContext.BaseDirectory, "data/scripts/lunr.interop.js");
+
+            return js.Run(scriptPath, "transform", documents).ToString();
+        }
+
+        private bool IsLunrSearchEnabled(FilePath file)
+        {
+            if (_config.SearchEngine != SearchEngineType.Lunr)
```
Comment (live1206) [inline src/docfx/build/search/SearchIndexBuilder.cs:63]: consider take `_config.SearchEngine != SearchEngineType.Lunr` as field, it does not depend on `file`

File: src/docfx/build/search/SearchIndexBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                return false;
+            }
+
+            return true;
+        }
+
+        [JsonObject(NamingStrategyType = typeof(CamelCaseNamingStrategy))]
+        private class LunrSearchItem
+        {
+            public string? Id { get; set; }
```
Comment (live1206) [inline src/docfx/build/search/SearchIndexBuilder.cs:80]: why is `Id`(`SiteUrl`) nullable?

File: src/docfx/build/page/BuildPage.cs [hunk 5]
```diff
@@ -176,6 +177,7 @@ private static SystemMetadata CreateSystemMetadata(ErrorBuilder errors, Context
 
             systemMetadata.SearchProduct = context.Config.Product;
             systemMetadata.SearchDocsetName = context.Config.Name;
+            systemMetadata.SearchEngine = context.Config.SearchEngine;
```
Comment (live1206) [inline src/docfx/build/page/BuildPage.cs:180]: what is `systemMetadata.SearchEngine` used for? seems like only `config.SearchEngine` is in use

Comment (yufeih) [reply src/docfx/build/page/BuildPage.cs:180]: It is used by template to what search UI to show based on the search engine type,

**Pull request merged**